### DISCUSSION TITLE
exiftool: no Keywords field in .MOV or .HEIC files?

### DISCUSSION THREAD

**Discussion by Brian-M-Holmes** [RhetTbull/osxphotos#1231] [Q&A] [answered]
I'm new to osxphotos and I think I have read most of the (excellent) documentation. This tool seems fantastic! Anyway, one of my goals is to embed Photos.app keywords, album names, and folder names into the exported files such that it would be possible to programatically re-create the logical albums, folders, and "keyword groups" at some future date. 

In my testing, I'm using the --exiftool and --keyword-template "{folder_album(>)} options on export. I then use the 'exiftool' command to display the exported image file's "EXIF fields" (as I'll call them; I'm not sure of the proper terminology). I notice for **most** images there exists an embedded EXIF field called "Keywords" which contains the expected Photos.app keywords and album names. Great! 

However, to my surprise there is no such embedded EXIF field called "Keywords" within my exported .MOV and .HEIC files, even if Photos.app has keywords assigned to them! I've seen no mention of this in the documentation.

Question (1): Is this expected behavior or a bug or a limitation of exiftool?

I did notice that (apparently) all types of exported files have the EXIF fields of "Subject" and "Tags List" which do appear to contain all the Photos.app keywords I was looking for. 

Question (2): Which "EXIF field" should I expect to be most reliable to contain an image's or movie's complete set of keywords?

Question (3): Also, does anyone whether the "Subject" and "Tags List" fields will always be exactly the same?

Any comments appreciated!

**Comment by Brian-M-Holmes** [comment]
Looking into exiftool more, I find a handy option when displaying the file's embedded metadata:

exiftool -groupNames myfile

shows me that the "Keywords" field is part of the IPTC group, while the "Subject" and "Tags List" fields are part of the XMP group. None of my .MOV or .HEIC files have any fields of the IPTC group. Perhaps IPTC fields are not allowed in those filetypes? 

While this would explain why my movie files don't include "Keywords", it is rather confusing for the osxphotos user, who might assume "Keywords" should contain the keywords assigned in Photos.app, even for movies or .heic photos.

Perhaps this is mostly a documentation issue?

**Reply by RhetTbull** [reply]
>  Perhaps IPTC fields are not allowed in those filetypes?

That's correct.

I'm not sure where in the docs I'd mention this. I don't want to provide a tutorial on photo metadata which is a very complex subject outside the scope of osxphotos. I'm open to suggestions or a pull request.

**Comment by RhetTbull** [comment, marked as answer]
Hi @Brian-M-Holmes Glad you found osxphotos useful. To answer your questions:

> Question (1): Is this expected behavior or a bug or a limitation of exiftool?

Yes, it's expected and it's a limitation of the actual file formats, not exiftool or osxphotos. osxphotos will use exiftool to write whichever metadata is supported by the file type. The `IPTC:Keywords` tag, which is what you're looking at as `Keywords` is not supported in HEIC or MOV files.

The `Subject` and `TagsList` are XMP fields which are supported. 

There are 3 types of metadata in image/video files: EXIF, IPTC, and XMP. osxphotos (and exiftool) support all three but not every file supports all three.

> Question (2): Which "EXIF field" should I expect to be most reliable to contain an image's or movie's complete set of keywords?

Every photo app works differently and prefers certain tags. Photo metadata is surprisingly not standardized. osxphotos attempts to use the most common fields in an effort to ensure that maximum metadata is preserved. The readme help text for `--exiftool` (try `osxphotos help export exiftool`) lists all the fields supported.

> Question (3): Also, does anyone whether the "Subject" and "Tags List" fields will always be exactly the same?

In the case of osxphotos, yes. I can make no such guarantees for other software.

**Comment by Brian-M-Holmes** [comment]
Thank you so much -- I think I'm understanding this better, after looking more into exiftool, and the documentation of the --exiftool and --sidecar options of osxphotos.

I plan to use osxphotos to "back up" my photos/videos in a simple filesystem organization by date (the --export-by-date is perfect for me). My goal here was to discover which embedded metadata fields would make it _possible_ for a script to re-create the groupings (albums, folders, and  smart albums based on keywords) that currently exist only in my Photos Library, and save those groupings in some other way. (For example, the script could create folders containing filesystem links to the relevant image and movie files.) I probably won't ever write such a script -- I just want to be reassured that it _could_ be done, if Apple's Photos apps were to somehow "go crazy" and totally mess up my library!

According to my testing so far, it seems that the XMP and EXIF groups will contain the metadata I desire and I can ignore the IPTC fields. 

Thank you for your help, and for authoring this impressive tool!

**Reply by RhetTbull** [reply]
> I plan to use osxphotos to "back up" my photos/videos in a simple filesystem organization by date.

That's exactly what I originally wrote osxphotos for.

> My goal here was to discover which embedded metadata fields would make it possible for a script to re-create the groupings.

The only metadata that isn't currently exported is the album though I plan to add that soon (#967) once I finish the current branch to add support for legacy iPhoto libraries. To preserve the album in the exported data you could use `--album-keyword` to include the album name as keyword template. You could also use `--keyword-template` to create a custom keyword. For example:
`--keyword-template "{folder_album(>)} ` in your original question will write something like "Folder>Subfolder>Album" which some apps like Lightroom classic can use to create hierarchical keywords.  You could also do something like `--keyword-template "ALBUM:{folder_album}"` so you know in the future this denotes an album. The template REPL is great for experimenting with these (`osxphotos template`)

With the exception of location, which is EXIF specific, all data is written to both XMP and IPTC fields. For example, `EXIF:ImageDescription`,  `XMP:Description`, and `IPTC:Caption-Abstract` all hold the description (caption).

**Reply by RhetTbull** [reply]
By the way, you can also create your own sidecar file using `--sidecar-template` which has the following help text. You could put whatever data you want into this to make it so your "recreation" script doesn't need to parse EXIF/IPTC/XMP metadata. Additionally, the osxphotos export database also has a full copy of all the metadata for each photo and if you keep the export database (`.osxphotos_export.db` in the export root) you could read that recreate the structure.


Create a custom sidecar file for each photo exported with user provided Mako template (MAKO_TEMPLATE_FILE). MAKO_TEMPLATE_FILE must be a valid Mako template (see https://www.makotemplates.org/). The template will passed the following variables: photo (PhotoInfo object for the photo being exported), sidecar_path (pathlib.Path object for the path to the sidecar being written), and photo_path (pathlib.Path object for the path to the exported photo. SIDECAR_FILENAME_TEMPLATE must be a valid template string (see Templating System in help) which will be rendered to generate the filename of the sidecar file. The {filepath} template variable may be used in the SIDECAR_FILENAME_TEMPLATE to refer to the filename of the photo being exported. OPTIONS is a comma-separated list of strings providing additional options to the template. Valid options are: write_skipped, strip_whitespace, strip_lines, skip_zero, catch_errors, none. write_skipped will cause the sidecar file to be written even if the photo is skipped during export. If write_skipped is not passed as an option, the sidecar file will not be written if the photo is skipped during export. strip_whitespace and strip_lines indicate whether or not to strip whitespace and blank lines, respectively, from the resulting sidecar file. skip_zero causes the sidecar file to be skipped if the rendered template is zero-length. catch_errors causes errors in the template to be caught and logged but not raised. Without catch_errors, osxphotos will abort the export if an error occurs in the template. For example, to create a sidecar file with extension .xmp using a template file named ‘sidecar.mako’ and write a sidecar for skipped photos and strip blank lines but not whitespace: –sidecar-template sidecar.mako ‘{filepath}.xmp’ write_skipped,strip_lines. To do the same but to drop the photo extension from the sidecar filename: –sidecar-template sidecar.mako ‘{filepath.parent}/{filepath.stem}.xmp’ write_skipped,strip_lines. If you are not passing any options, you must pass ‘none’ as the last argument to –sidecar-template: –sidecar-template sidecar.mako ‘{filepath}.xmp’ none. For an example Mako file see https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/custom_sidecar.mako

**Reply by RhetTbull** [reply]
If you run the following command, you'll get a JSON dump of everything osxphotos knows about an exported file that could be used to recreate the original path, keywords, etc. 

`osxphotos exportdb /export/path --info /export/path/IMG_1234.HEIC`

The code used to do this if you wanted to do so in your own python script is:

https://github.com/RhetTbull/osxphotos/blob/ada0e92f69f5730fb42cc3f8b69a0290450f3e35/osxphotos/cli/exportdb.py#L344-L357
